## Trabalho Individual I

Marco Delgado Esperança, nº 110451, CDB1

## Problema 1: Minimização da Soma dos Desvios

## a)

Atendendo que se pretende obter soluções que se aproximem do valor alvo de cada
meta, então o natural é minimizar a soma dos valores dos desvios (penalizações), ou seja:
 

## $\min P = 5*\frac{d_1^{-}}{125} + \frac{10}{5}*\frac{d_2^{-}}{60} + \frac{4}{5}*\frac{d_2^{+}}{60} + 5*\frac{d_3^{+}}{55}$



##   s.a. $x_1 \leq 6$ 
##          $x_2 \geq 2$ 
##          $x_3 \geq 1$ 
##          $12x_1 + 9x_2 + 5x_3 \geq 125$
##          $5x_1 + 2x_2 + 4x_3 = 60$
##          $5x_1 + 5x_2 + 8x_3 \leq 55$
##          $x_1, x_2, x_3 \geq 0$
##          $d_1^-, d_2^-, d_2^+, d_3^+ \geq 0$

## b)

## Modelo

In [102]:
import pandas as pd
import numpy as np
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, LpVariable
from pulp import GLPK

#
# Create the model
model = LpProblem(name="Projeto_individual_1_b", sense= LpMinimize)

#
# Initialize the decision variables
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2,4)}


#
# Add the constraints to the model
model += (x[1] <= 6, "produção D1")
model += (x[2] >= 2, "produção D2")
model += (x[3] >= 1, "produção D3")
model += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1] >= 125, "Lucro a longo prazo")
model += (5 * x[1] + 2 * x[2] + 4 * x[3] - dM[2] + dm[2]== 60, "Mão de obra")
model += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "Investimento")
#
# Add the objective function to the model
vector = np.array([125, 60, 55])
peso = np.array([5,4/5,10/5,5])
obj_func = (peso[0]/vector[0]) * dm[1] + (peso[1]/vector[1]) * dM[2] + (peso[2]/vector[1]) * dm[2]+ (peso[3]/vector[2]) * dM[3]
model += obj_func

# 
# Formulação (visualização do modelo matemático de forma a conferir os dados)
model

Projeto_individual_1_b:
MINIMIZE
0.013333333333333334*dM2 + 0.09090909090909091*dM3 + 0.04*dm1 + 0.03333333333333333*dm2 + 0.0
SUBJECT TO
produção_D1: x1 <= 6

produção_D2: x2 >= 2

produção_D3: x3 >= 1

Lucro_a_longo_prazo: dm1 + 12 x1 + 9 x2 + 5 x3 >= 125

Mão_de_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

Investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dM2 Continuous
dM3 Continuous
dm1 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

## Resolução do problema e Resultados

In [103]:
# Solve the problem
#
# Consideram-se três formas de resolver um problema de Programação Linear: 
# Forma 1: status = model.solve() 
# Forma 2: status = model.solve(solver=GLPK(msg=True))
# Forma 3: status = model.solve(GLPK(msg=True, options=['--ranges', 'sensitivity.txt']))
# A Forma 3 gera o relatório de análise de sensibilidade.
#
status = model.solve()

#
# Outputs
#
# Status (LpStatusOptimal=1; LpStatusNotSolved=0; LpStatusInfeasible=-1; LpStatusUnbounded=-2; LpStatusUndefined=-3)
#
# Valor óptimo de problema
model.objective.value()
print(f"objective: {model.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das variáveis de desvio
#
for name, constraint in model.constraints.items():
    print (f"{name}: {constraint.value()}") 


Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/marcoesperanca/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/a870a9ae72e543b38254456a170d96a3-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/a870a9ae72e543b38254456a170d96a3-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 11 COLUMNS
At line 32 RHS
At line 39 BOUNDS
At line 40 ENDATA
Problem MODEL has 6 rows, 7 columns and 16 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 3 (-3) rows, 7 (0) columns and 13 (-3) elements
0  Obj 0 Primal inf 18.899998 (2)
5  Obj 1.336
Optimal - objective value 1.336
After Postsolve, objective 1.336, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 1.336 - 5 iterations time 0.002, Presolve 0.00
Option for printingOptions chang

<br><br>

## Interpretação dos Resultados

### Valor ótimo: 
P* = 1.3359. Este valor ótimo é difícil de interpretar pois representa a soma de desvios de diferentes naturezas. No entanto, pretende-se um valor baixo, pois, corresponde à soma das penalizações, que podemos concluir que foi atingido.

### Solução Ótima

X1 = 6.0, ou seja, são produzidos 6000 kg de doces D1, que é igual ao valor máximo de produção deste doce permitido (produção_D1 = 0.0)<br>
X2 = 3.4, ou seja, são produzidos 3400 kg de doces D2, que está 1400 kg acima do valor mínimo exigido (produção_D2 = 1.4)<br>
X3 = 1.0, ou seja, são produzidos 1000 kg de doces D3, que é igual ao valor mínimo de produção solicitado (produção_D3 = 0.0)<br><br>

dm1 = 17.4 > 0, ou seja, o lucro a longo prazo é de 125000 - 17400, isto é, 107600 euros, ou seja, não é alcançado o valor da meta do lucro do longo prazo, pois, encontra-se abaixo 17400 euros (dm1=17.4) do valor mínimo exigido (lucro de pelo menos 125 mil euros).
<br><br>
dm2 = 19.2 > 0, ou seja, a mão de obra é de 60-19.2 (dm2=19.2), isto é, 40.8, o que corresponde a cerca de  40 empregados, não respeitando a meta do número de funcionários ser 60.
<br><br>
dM2 = 0.0, ou seja, não há qualquer penalização de por cada 5 trabalhadores, 4 pontos por ultrapassar o valor de referência.
<br><br>
dM3 = 0.0, ou seja, não há qualquer penalização de 5 pontos por cada milhar de euros acima do nível de aspiração do investimento de capital.

<br><br>

## c)

## Alternativa 1

O primeiro método alternativo que será apresentado para o problema é definido por outro tipo de função objetivo, designado por **Objetivo MiniMax**, a qual é utilizada quando se pretende minimizar o desvio máximo de qualquer meta.

## Temos que: $P_1^-=5, P_2^- = \frac{10}{5}, P_2^+ = \frac{4}{5}, P_3^+ = 5$

## $\min \max Z = \left\{ 5 * \frac{d_1^{-}}{125}, \frac{10}{5} * \frac{d_2^{-}}{60}, \frac{4}{5} * \frac{d_2^{+}}{60}, 5 * \frac{d_3^{+}}{55} \right\}$

## Função objetivo: $\min Z = Q$, onde $Q$ é a variável **MiniMax**.

##   s.a. $x_1 \leq 6$
##          $x_2 \geq 2$
##          $x_3 \geq 1$
##          $12x_1 + 9x_2 + 5x_3 + d_1^-\geq 125$
##          $5x_1 + 2x_2 + 4x_3  + d_2^+ - d_2^- = 60$
##          $5x_1 + 5x_2 + 8x_3 - d_3^+ \leq 55$
##          $\begin{equation} P_{1}^{-} * \frac{d_{1}^{-}}{125} \leq Q\end{equation}$
##          $\begin{equation} P_{2}^{-} * \frac{d_{2}^{-}}{60} \leq Q\end{equation}$
##          $\begin{equation} P_{2}^{+} * \frac{d_{2}^{+}}{60} \leq Q\end{equation}$
##          $\begin{equation} P_{3}^{+} * \frac{d_{3}^{+}}{55} \leq Q\end{equation}$
##          $x_1, x_2, x_3 \geq 0$
##          $d_1^-, d_2^-, d_2^+, d_3^+ \geq 0$
##          $Q \geq 0$

## Modelo com os mesmos pesos, mas usando MiniMax

In [86]:
import pandas as pd
import numpy as np
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, LpVariable
from pulp import GLPK

## Create the model
model = LpProblem(name="Projeto_individual_1_c_1", sense= LpMinimize)


#
# Initialize the decision variables
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2,4)}
Q = LpVariable("Q", lowBound=0)


#
# Add the constraints to the model
model += (x[1] <= 6, "produção D1")
model += (x[2] >= 2, "produção D2")
model += (x[3] >= 1, "produção D3")
model += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1] >= 125, "Lucro a longo prazo")
model += (5 * x[1] + 2 * x[2] + 4 * x[3] - dM[2] + dm[2]== 60, "Mão de obra")
model += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "Investimento")

## Add the objective function to the model
vector = np.array([125, 5*60, 55])
peso = np.array([5,4,10,5])


model += ((peso[0]/vector[0]) * dm[1] <= Q, "Restricao_dm1_lucro" )
model += ((peso[1]/vector[1]) * dM[2] <= Q, "Restricao_dM2_trabalhadores" )
model += ((peso[2]/vector[1]) * dm[2] <= Q, "Restricao_dm2_trabalhadores" )
model += ((peso[3]/vector[2]) * dM[3] <= Q, "Restrição_dM3_investimento" )


obj_func = Q
model += obj_func

## Formulação (visualização do modelo matemático de forma a conferir os dados)
model

Projeto_individual_1_c_1:
MINIMIZE
1*Q + 0
SUBJECT TO
produção_D1: x1 <= 6

produção_D2: x2 >= 2

produção_D3: x3 >= 1

Lucro_a_longo_prazo: dm1 + 12 x1 + 9 x2 + 5 x3 >= 125

Mão_de_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

Investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

Restricao_dm1_lucro: - Q + 0.04 dm1 <= 0

Restricao_dM2_trabalhadores: - Q + 0.0133333333333 dM2 <= 0

Restricao_dm2_trabalhadores: - Q + 0.0333333333333 dm2 <= 0

Restrição_dM3_investimento: - Q + 0.0909090909091 dM3 <= 0

VARIABLES
Q Continuous
dM2 Continuous
dM3 Continuous
dm1 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

## Resolução de problemas e resultados

In [87]:
# Solve the problem
#
# Consideram-se três formas de resolver um problema de Programação Linear: 
# Forma 1: status = model.solve() 
# Forma 2: status = model.solve(solver=GLPK(msg=True))
# Forma 3: status = model.solve(GLPK(msg=True, options=['--ranges', 'sensitivity.txt']))
# A Forma 3 gera o relatório de análise de sensibilidade.
#
status = model.solve()

#
# Outputs
#
# Status (LpStatusOptimal=1; LpStatusNotSolved=0; LpStatusInfeasible=-1; LpStatusUnbounded=-2; LpStatusUndefined=-3)
#
# Valor óptimo de problema
model.objective.value()
print(f"objective: {model.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
print(f"{Q}: {Q.value()}")
#
# Valores das variáveis de desvio
#
for name, constraint in model.constraints.items():
    print (f"{name}: {constraint.value()}") 

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/marcoesperanca/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/45465f45f8164369ba1de0713d8a0708-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/45465f45f8164369ba1de0713d8a0708-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 15 COLUMNS
At line 41 RHS
At line 52 BOUNDS
At line 53 ENDATA
Problem MODEL has 10 rows, 8 columns and 24 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 5 (-5) rows, 6 (-2) columns and 17 (-7) elements
0  Obj 0 Primal inf 66.29457 (2)
6  Obj 0.54120267
Optimal - objective value 0.54120267
After Postsolve, objective 0.54120267, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 0.5412026726 - 6 iterations time 0.002, Presolve 0.00
Option fo

## Interpretação dos Resultados

### Valor ótimo: 
Z* = 0.54120267. Este valor ótimo é difícil de interpretar pois representa a soma de desvios de diferentes naturezas. No entanto, pretende-se um valor baixo, pois, pretende-se miniimzar o desvio máximo de todas as metas, o que podemos concluir que foi atingido.

### Solução Ótima

X1 = 6.0, ou seja, são produzidos 6000 kg de doces D1, que é igual ao valor máximo de produção deste doce permitido (produção_D1 = 0.0)<br>
X2 = 3.4253898, ou seja, são produzidos cerca de 3425 kg de doces D2, que está cerca de 1400 kg acima do valor mínimo exigido (produção_D2 = 1.4253898)<br>
X3 = 1.7282851, ou seja, são produzidos cerca de 1728 kg de doces D3, que é cerca de 0.7 kg superior ao valor mínimo de produção solicitado (produção_D3 = 0.7282850999999999)<br><br>

dm1 = 17.4 > 0, ou seja, o lucro a longo prazo é de 125000 - 13.530067, isto é, aproximadamente, 124986 euros, ou seja, não é alcançado o valor da meta do lucro do longo prazo, pois, encontra-se abaixo cerca de 13530 euros (dm1=13.530067) do valor mínimo exigido (lucro de pelo menos 125 mil euros).
<br><br>
dm2 = 16.23608 > 0, ou seja, a mão de obra é de 60-16.23608 (dm2=16.23608), isto é, 43.76, o que corresponde a cerca de 43 empregados, não respeitando a meta do número de funcionários ser 60.
<br><br>
dM2 = 0.0, ou seja, não há qualquer penalização de por cada 5 trabalhadores, 4 pontos por ultrapassar o valor de referência.
<br><br>
dM3 = 5.9532294, ou seja, há penalização de 5 pontos por cada milhar de euros acima do nível de aspiração do investimento de capital.

<br><br>

## Alternativa 2.1

## Mesma formulação da b) sem MinMax, mas com pesos diferentes

A formulação é a mesma da b), mas com pesos diferentes.

In [88]:
import pandas as pd
import numpy as np
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, LpVariable
from pulp import GLPK

#
# Create the model
model = LpProblem(name="Projeto_individual_c_2", sense= LpMinimize)

#
# Initialize the decision variables
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2,4)}


#
# Add the constraints to the model
model += (x[1] <= 6, "produção D1")
model += (x[2] >= 2, "produção D2")
model += (x[3] >= 1, "produção D3")
model += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1] >= 125, "Lucro a longo prazo")
model += (5 * x[1] + 2 * x[2] + 4 * x[3] - dM[2] + dm[2]== 60, "Mão de obra")
model += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "Investimento")
#
# Add the objective function to the model
vector = np.array([125, 60, 55])


peso = np.array([3,1,2,1])



obj_func = (peso[0]/vector[0]) * dm[1] + (peso[1]/vector[1]) * dM[2] + (peso[2]/vector[1]) * dm[2] + (peso[3]/vector[2]) * dM[3]
model += obj_func

# 
# Formulação (visualização do modelo matemático de forma a conferir os dados)
model

Projeto_individual_c_2:
MINIMIZE
0.016666666666666666*dM2 + 0.01818181818181818*dM3 + 0.024*dm1 + 0.03333333333333333*dm2 + 0.0
SUBJECT TO
produção_D1: x1 <= 6

produção_D2: x2 >= 2

produção_D3: x3 >= 1

Lucro_a_longo_prazo: dm1 + 12 x1 + 9 x2 + 5 x3 >= 125

Mão_de_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

Investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dM2 Continuous
dM3 Continuous
dm1 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [89]:
# Solve the problem
#
# Consideram-se três formas de resolver um problema de Programação Linear: 
# Forma 1: status = model.solve() 
# Forma 2: status = model.solve(solver=GLPK(msg=True))
# Forma 3: status = model.solve(GLPK(msg=True, options=['--ranges', 'sensitivity.txt']))
# A Forma 3 gera o relatório de análise de sensibilidade.
#
status = model.solve()

#
# Outputs
#
# Status (LpStatusOptimal=1; LpStatusNotSolved=0; LpStatusInfeasible=-1; LpStatusUnbounded=-2; LpStatusUndefined=-3)
#
# Valor óptimo de problema
model.objective.value()
print(f"objective: {model.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das variáveis de desvio
#
for name, constraint in model.constraints.items():
    print (f"{name}: {constraint.value()}") 

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/marcoesperanca/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/5c26ce32f9ad4ff5948c15d828fe2856-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/5c26ce32f9ad4ff5948c15d828fe2856-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 11 COLUMNS
At line 32 RHS
At line 39 BOUNDS
At line 40 ENDATA
Problem MODEL has 6 rows, 7 columns and 16 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 3 (-3) rows, 7 (0) columns and 13 (-3) elements
0  Obj 0 Primal inf 18.899998 (2)
5  Obj 0.67972028
Optimal - objective value 0.67972028
After Postsolve, objective 0.67972028, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 0.6797202797 - 5 iterations time 0.002, Presolve 0.00
Option for

## Alternativa 2.2

In [100]:
import pandas as pd
import numpy as np
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, LpVariable
from pulp import GLPK

#
# Create the model
model = LpProblem(name="Projeto_individual_c_3", sense= LpMinimize)


# Initialize the decision variables
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1, 4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1, 3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2,4)}


#
# Add the constraints to the model
model += (x[1] <= 6, "produção D1")
model += (x[2] >= 2, "produção D2")
model += (x[3] >= 1, "produção D3")
model += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1] >= 125, "Lucro a longo prazo")
model += (5 * x[1] + 2 * x[2] + 4 * x[3] - dM[2] + dm[2]== 60, "Mão de obra")
model += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "Investimento")
#
# Add the objective function to the model
vector = np.array([125, 60, 55])


peso = np.array([3,2,2,3])



obj_func = (peso[0]/vector[0]) * dm[1] + (peso[1]/vector[1]) * dM[2] + (peso[2]/vector[1]) * dm[2] + (peso[3]/vector[2]) * dM[3]
model += obj_func

# 
# Formulação (visualização do modelo matemático de forma a conferir os dados)
model

Projeto_individual_c_3:
MINIMIZE
0.03333333333333333*dM2 + 0.05454545454545454*dM3 + 0.024*dm1 + 0.03333333333333333*dm2 + 0.0
SUBJECT TO
produção_D1: x1 <= 6

produção_D2: x2 >= 2

produção_D3: x3 >= 1

Lucro_a_longo_prazo: dm1 + 12 x1 + 9 x2 + 5 x3 >= 125

Mão_de_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

Investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dM2 Continuous
dM3 Continuous
dm1 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [101]:
# Solve the problem
#
# Consideram-se três formas de resolver um problema de Programação Linear: 
# Forma 1: status = model.solve() 
# Forma 2: status = model.solve(solver=GLPK(msg=True))
# Forma 3: status = model.solve(GLPK(msg=True, options=['--ranges', 'sensitivity.txt']))
# A Forma 3 gera o relatório de análise de sensibilidade.
#
status = model.solve()

#
# Outputs
#
# Status (LpStatusOptimal=1; LpStatusNotSolved=0; LpStatusInfeasible=-1; LpStatusUnbounded=-2; LpStatusUndefined=-3)
#
# Valor óptimo de problema
model.objective.value()
print(f"objective: {model.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das variáveis de desvio
#
for name, constraint in model.constraints.items():
    print (f"{name}: {constraint.value()}") 

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/marcoesperanca/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/280acddbb80c4518bafa3059d64587f1-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/280acddbb80c4518bafa3059d64587f1-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 11 COLUMNS
At line 32 RHS
At line 39 BOUNDS
At line 40 ENDATA
Problem MODEL has 6 rows, 7 columns and 16 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 3 (-3) rows, 7 (0) columns and 13 (-3) elements
0  Obj 0 Primal inf 18.899998 (2)
5  Obj 1.0383838
Optimal - objective value 1.0383838
After Postsolve, objective 1.0383838, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 1.038383838 - 5 iterations time 0.002, Presolve 0.00
Option for pri

###  Interpretação dos resultados desta alínea e das restantes (ver relatório)

## e)

## Progração por Metas Preemptiva

A direção da Confeitaria reavaliou as prioridades das três metas e decidiu dar uma prioridade muito elevada à meta de mão de obra, bem como à meta de investimento de capital. A meta de angariar mais de 55 milhões de euros para investimento de capital foi considerada extremamente difícil de ser alcançada. <br>
Como se estabelece uma hierarquia de níveis de prioridade, trata-se de **Programação por Metas Preemptiva**.

![prioridade_metas](prioridade_metas.png)

## Formulação do problema
<br>

## $Lex \min Z = \left\{\frac{P_3^+ * d_3^+}{55} + \frac{P_2^+ * d_2^+}{60} + \frac{P_2^- * d_2^-}{60}, \frac{P_1^- * d_1^-}{125} \right\}$
<br>

### **Primeiro nível**

## $\min P = P_2^-*\frac{d_2^{-}}{60} + P_2^+*\frac{d_2^{+}}{60} + P_3^+*\frac{d_3^{+}}{55}$


##   s.a. $x_1 \leq 6$
##          $x_2 \geq 2$
##          $x_3 \geq 1$
##          $5x_1 + 2x_2 + 4x_3 - d_2^+ + d_2^- = 60$
##          $5x_1 + 5x_2 + 8x_3 - d_3^+ \leq 55$
##          $x_1, x_2, x_3 \geq 0$
##          $d_2^-, d_2^+, d_3^+ \geq 0$

## Pesos iniciais

### Primeiro nível (primeira tentativa)

In [1]:
import pandas as pd
import numpy as np
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, LpVariable
from pulp import GLPK

#
# Create the model
model_ppn = LpProblem(name="Projeto_individual_1_e_pn", sense= LpMinimize)

#
# Initialize the decision variables
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1,4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(2,3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2,4)}

vector = np.array([60,55])
peso = np.array([10/5,4/5,5])

#
# Add the constraints to the model
model_ppn += (x[1] <= 6, "produção D1")
model_ppn += (x[2] >= 2, "produção D2")
model_ppn += (x[3] >= 1, "produção D3")
model_ppn += (5 * x[1] + 2 * x[2] + 4 * x[3] - dM[2] + dm[2] == 60, "Mão de obra")
model_ppn += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "Investimento")
#
# Add the objective function to the model
obj_func = (peso[0]/vector[0]) * dm[2] + (peso[1]/vector[0]) * dM[2] + (peso[2]/vector[1]) * dM[3]
model_ppn += obj_func

# 
# Formulação (visualização do modelo matemático de forma a conferir os dados)
model_ppn

Projeto_individual_1_e_pn:
MINIMIZE
0.013333333333333334*dM2 + 0.09090909090909091*dM3 + 0.03333333333333333*dm2 + 0.0
SUBJECT TO
produção_D1: x1 <= 6

produção_D2: x2 >= 2

produção_D3: x3 >= 1

Mão_de_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

Investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dM2 Continuous
dM3 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [2]:
# Solve the problem
#
# Consideram-se três formas de resolver um problema de Programação Linear: 
# Forma 1: status = model.solve() 
# Forma 2: status = model.solve(solver=GLPK(msg=True))
# Forma 3: status = model.solve(GLPK(msg=True, options=['--ranges', 'sensitivity.txt']))
# A Forma 3 gera o relatório de análise de sensibilidade.
#
status = model_ppn.solve()

#
# Outputs
#
# Status (LpStatusOptimal=1; LpStatusNotSolved=0; LpStatusInfeasible=-1; LpStatusUnbounded=-2; LpStatusUndefined=-3)
#
# Valor óptimo de problema
model_ppn.objective.value()
print(f"objective: {model_ppn.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das variáveis de desvio
#
for name, constraint in model_ppn.constraints.items():
    print (f"{name}: {constraint.value()}") 

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/marcoesperanca/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/78059453014c4572ab02ffff461ce6ac-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/78059453014c4572ab02ffff461ce6ac-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 10 COLUMNS
At line 26 RHS
At line 32 BOUNDS
At line 33 ENDATA
Problem MODEL has 5 rows, 6 columns and 12 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 2 (-3) rows, 6 (0) columns and 9 (-3) elements
0  Obj 0 Primal inf 10.399999 (1)
3  Obj 0.61666667
Optimal - objective value 0.61666667
After Postsolve, objective 0.61666667, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 0.6166666667 - 3 iterations time 0.002, Presolve 0.00
Option for 

Número ideal de trabalhadores: 60-18.5 = 41.5 (41 trabalhadores)

### Primeiro nível (segunda tentativa)

#### Reformulação do problema

## $\min P = P_2^-*\frac{d_2^{-}}{60} + P_2^+*\frac{d_2^{+}}{60} + P_3^+*\frac{d_3^{+}}{55}$


##   s.a. $x_1 \leq 6$
##          $x_2 \geq 2$
##          $x_3 \geq 1$
##          $5x_1 + 2x_2 + 4x_3 - d_2^+ + d_2^- = 41$
##          $5x_1 + 5x_2 + 8x_3 - d_3^+ \leq 55$
##          $x_1, x_2, x_3 \geq 0$
##          $d_2^-, d_2^+, d_3^+ \geq 0$

In [3]:
import pandas as pd
import numpy as np
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, LpVariable
from pulp import GLPK

#
# Create the model
model_ppn = LpProblem(name="Projeto_individual_1_e_pn", sense= LpMinimize)

#
# Initialize the decision variables
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1,4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(2,3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2,4)}

vector = np.array([60,55])
peso = np.array([10/5,4/5,5])

#
# Add the constraints to the model
model_ppn += (x[1] <= 6, "produção D1")
model_ppn += (x[2] >= 2, "produção D2")
model_ppn += (x[3] >= 1, "produção D3")
model_ppn += (5 * x[1] + 2 * x[2] + 4 * x[3] - dM[2] + dm[2] == 41, "Mão de obra")
model_ppn += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "Investimento")
#
# Add the objective function to the model
obj_func = (peso[0]/vector[0]) * dm[2] + (peso[1]/vector[0]) * dM[2] + (peso[2]/vector[1]) * dM[3]
model_ppn += obj_func

# 
# Formulação (visualização do modelo matemático de forma a conferir os dados)
model_ppn

Projeto_individual_1_e_pn:
MINIMIZE
0.013333333333333334*dM2 + 0.09090909090909091*dM3 + 0.03333333333333333*dm2 + 0.0
SUBJECT TO
produção_D1: x1 <= 6

produção_D2: x2 >= 2

produção_D3: x3 >= 1

Mão_de_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 41

Investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dM2 Continuous
dM3 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [4]:
# Solve the problem
#
# Consideram-se três formas de resolver um problema de Programação Linear: 
# Forma 1: status = model.solve() 
# Forma 2: status = model.solve(solver=GLPK(msg=True))
# Forma 3: status = model.solve(GLPK(msg=True, options=['--ranges', 'sensitivity.txt']))
# A Forma 3 gera o relatório de análise de sensibilidade.
#
status = model_ppn.solve()

#
# Outputs
#
# Status (LpStatusOptimal=1; LpStatusNotSolved=0; LpStatusInfeasible=-1; LpStatusUnbounded=-2; LpStatusUndefined=-3)
#
# Valor óptimo de problema
model_ppn.objective.value()
print(f"objective: {model_ppn.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das variáveis de desvio
#
for name, constraint in model_ppn.constraints.items():
    print (f"{name}: {constraint.value()}") 

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/marcoesperanca/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/d6f70a3bb0ff4f05ba731db5741f38da-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/d6f70a3bb0ff4f05ba731db5741f38da-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 10 COLUMNS
At line 26 RHS
At line 32 BOUNDS
At line 33 ENDATA
Problem MODEL has 5 rows, 6 columns and 12 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 2 (-3) rows, 6 (0) columns and 9 (-3) elements
0  Obj 0 Primal inf 6.599999 (1)
2  Obj 1.4e-18
Optimal - objective value 0
After Postsolve, objective 0, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 0 - 2 iterations time 0.002, Presolve 0.00
Option for printingOptions changed from norm

### Segundo nível (primeira tentativa)

In [35]:
import pandas as pd
import numpy as np
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, LpVariable
from pulp import GLPK

#
# Create the model
model_psn = LpProblem(name="Projeto_individual_1_e_sn", sense= LpMinimize)

#
# Initialize the decision variables
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1,4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1,2)}


vector = np.array([125])
peso = np.array([5])


# Add the constraints to the model
model_psn += (x[1] <= 6, "produção D1")
model_psn += (x[2] >= 2, "produção D2")
model_psn += (x[3] >= 1, "produção D3")
model_psn += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1] >= 125, "Lucro a longo prazo")
model_psn += (5 * x[1] + 2 * x[2] + 4 * x[3]  == 41, "Mão de obra")
model_psn += (5 * x[1] + 5 * x[2] + 8 * x[3] <= 55, "Investimento")
#
# Add the objective function to the model
obj_func = (peso[0]/vector[0]) * dm[1]
model_psn += obj_func

# 
# Formulação (visualização do modelo matemático de forma a conferir os dados)
model_psn

Projeto_individual_1_e_sn:
MINIMIZE
0.04*dm1 + 0.0
SUBJECT TO
produção_D1: x1 <= 6

produção_D2: x2 >= 2

produção_D3: x3 >= 1

Lucro_a_longo_prazo: dm1 + 12 x1 + 9 x2 + 5 x3 >= 125

Mão_de_obra: 5 x1 + 2 x2 + 4 x3 = 41

Investimento: 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dm1 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [36]:
# Solve the problem
#
# Consideram-se três formas de resolver um problema de Programação Linear: 
# Forma 1: status = model.solve() 
# Forma 2: status = model.solve(solver=GLPK(msg=True))
# Forma 3: status = model.solve(GLPK(msg=True, options=['--ranges', 'sensitivity.txt']))
# A Forma 3 gera o relatório de análise de sensibilidade.
#
status = model_psn.solve()

#
# Outputs
#
# Status (LpStatusOptimal=1; LpStatusNotSolved=0; LpStatusInfeasible=-1; LpStatusUnbounded=-2; LpStatusUndefined=-3)
#
# Valor óptimo de problema
model_psn.objective.value()
print(f"objective: {model_psn.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das variáveis de desvio
#
for name, constraint in model_psn.constraints.items():
    print (f"{name}: {constraint.value()}") 

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/marcoesperanca/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/6c67cabd1c654319b4a53300c0066f07-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/6c67cabd1c654319b4a53300c0066f07-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 11 COLUMNS
At line 26 RHS
At line 33 BOUNDS
At line 34 ENDATA
Problem MODEL has 6 rows, 4 columns and 13 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 3 (-3) rows, 4 (0) columns and 10 (-3) elements
0  Obj 0 Primal inf 15.099998 (2)
4  Obj 0.79
Optimal - objective value 0.79
After Postsolve, objective 0.79, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 0.79 - 4 iterations time 0.002, Presolve 0.00
Option for printingOptions changed f

<br>

## Segundo nível (segunda tentativa)

### Reformulação do problema

## $\min P = P_1^-*\frac{d_1^{-}}{125}$


##   s.a. $x_1 \leq 6$
##          $x_2 \geq 2$
##          $x_3 \geq 1$
##          $12x_1 + 9x_2 + 5x_3 + d_1^-\geq 105.25$
##          $5x_1 + 2x_2 + 4x_3 = 41$
##          $5x_1 + 5x_2 + 8x_3  \leq 55$
##          $x_1, x_2, x_3 \geq 0$
##          $d_1^- \geq 0$

In [33]:
import pandas as pd
import numpy as np
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, LpVariable
from pulp import GLPK

#
# Create the model
model_psn = LpProblem(name="Projeto_individual_1_e_sn", sense= LpMinimize)

#
# Initialize the decision variables
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1,4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1,2)}


vector = np.array([125])
peso = np.array([5])


# Add the constraints to the model
model_psn += (x[1] <= 6, "produção D1")
model_psn += (x[2] >= 2, "produção D2")
model_psn += (x[3] >= 1, "produção D3")
model_psn += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1] >= 105.25, "Lucro a longo prazo")
model_psn += (5 * x[1] + 2 * x[2] + 4 * x[3]  == 41, "Mão de obra")
model_psn += (5 * x[1] + 5 * x[2] + 8 * x[3] <= 55, "Investimento")
#
# Add the objective function to the model
obj_func = (peso[0]/vector[0]) * dm[1]
model_psn += obj_func

# 
# Formulação (visualização do modelo matemático de forma a conferir os dados)
model_psn

Projeto_individual_1_e_sn:
MINIMIZE
0.04*dm1 + 0.0
SUBJECT TO
produção_D1: x1 <= 6

produção_D2: x2 >= 2

produção_D3: x3 >= 1

Lucro_a_longo_prazo: dm1 + 12 x1 + 9 x2 + 5 x3 >= 105.25

Mão_de_obra: 5 x1 + 2 x2 + 4 x3 = 41

Investimento: 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dm1 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [34]:
# Solve the problem
#
# Consideram-se três formas de resolver um problema de Programação Linear: 
# Forma 1: status = model.solve() 
# Forma 2: status = model.solve(solver=GLPK(msg=True))
# Forma 3: status = model.solve(GLPK(msg=True, options=['--ranges', 'sensitivity.txt']))
# A Forma 3 gera o relatório de análise de sensibilidade.
#
status = model_psn.solve()

#
# Outputs
#
# Status (LpStatusOptimal=1; LpStatusNotSolved=0; LpStatusInfeasible=-1; LpStatusUnbounded=-2; LpStatusUndefined=-3)
#
# Valor óptimo de problema
model_psn.objective.value()
print(f"objective: {model_psn.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das variáveis de desvio
#
for name, constraint in model_psn.constraints.items():
    print (f"{name}: {constraint.value()}") 

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/marcoesperanca/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/c9e4b452a7694fe9b3af2e6a8a20ecc6-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/c9e4b452a7694fe9b3af2e6a8a20ecc6-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 11 COLUMNS
At line 26 RHS
At line 33 BOUNDS
At line 34 ENDATA
Problem MODEL has 6 rows, 4 columns and 13 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 3 (-3) rows, 4 (0) columns and 10 (-3) elements
0  Obj 0 Primal inf 13.454165 (2)
3  Obj 1.0318452e-17
Optimal - objective value 0
After Postsolve, objective 0, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 0 - 3 iterations time 0.002, Presolve 0.00
Option for printingOptions changed f

## $P_1^- = 5, P_2^- = 5, P_2^+ = 3, P_3^+ = 2$

### Primeiro nível (primeira tentativa)

In [53]:
import pandas as pd
import numpy as np
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, LpVariable
from pulp import GLPK

#
# Create the model
model_ppn = LpProblem(name="Projeto_individual_1_e_pn", sense= LpMinimize)

#
# Initialize the decision variables
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1,4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(2,3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2,4)}

vector = np.array([60,55])
peso = np.array([5,3,2])

#
# Add the constraints to the model
model_ppn += (x[1] <= 6, "produção D1")
model_ppn += (x[2] >= 2, "produção D2")
model_ppn += (x[3] >= 1, "produção D3")
model_ppn += (5 * x[1] + 2 * x[2] + 4 * x[3] - dM[2] + dm[2] == 60, "Mão de obra")
model_ppn += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 55, "Investimento")
#
# Add the objective function to the model
obj_func = (peso[0]/vector[0]) * dm[2] + (peso[1]/vector[0]) * dM[2] + (peso[2]/vector[1]) * dM[3]
model_ppn += obj_func

# 
# Formulação (visualização do modelo matemático de forma a conferir os dados)
model_ppn

Projeto_individual_1_e_pn:
MINIMIZE
0.05*dM2 + 0.03636363636363636*dM3 + 0.08333333333333333*dm2 + 0.0
SUBJECT TO
produção_D1: x1 <= 6

produção_D2: x2 >= 2

produção_D3: x3 >= 1

Mão_de_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

Investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 55

VARIABLES
dM2 Continuous
dM3 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [54]:
# Solve the problem
#
# Consideram-se três formas de resolver um problema de Programação Linear: 
# Forma 1: status = model.solve() 
# Forma 2: status = model.solve(solver=GLPK(msg=True))
# Forma 3: status = model.solve(GLPK(msg=True, options=['--ranges', 'sensitivity.txt']))
# A Forma 3 gera o relatório de análise de sensibilidade.
#
status = model_ppn.solve()

#
# Outputs
#
# Status (LpStatusOptimal=1; LpStatusNotSolved=0; LpStatusInfeasible=-1; LpStatusUnbounded=-2; LpStatusUndefined=-3)
#
# Valor óptimo de problema
model_ppn.objective.value()
print(f"objective: {model_ppn.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das variáveis de desvio
#
for name, constraint in model_ppn.constraints.items():
    print (f"{name}: {constraint.value()}") 

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/marcoesperanca/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/cea5fc39f3eb49c6b2ea491db189f1b5-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/cea5fc39f3eb49c6b2ea491db189f1b5-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 10 COLUMNS
At line 26 RHS
At line 32 BOUNDS
At line 33 ENDATA
Problem MODEL has 5 rows, 6 columns and 12 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 2 (-3) rows, 6 (0) columns and 9 (-3) elements
0  Obj 0 Primal inf 10.399999 (1)
3  Obj 1.3454545
Optimal - objective value 1.3454545
After Postsolve, objective 1.3454545, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 1.345454545 - 3 iterations time 0.002, Presolve 0.00
Option for prin

### Reformulação do problema

## $\min P = P_2^-*\frac{d_2^{-}}{60} + P_2^+*\frac{d_2^{+}}{60} + P_3^+*\frac{d_3^{+}}{55}$


##   s.a. $x_1 \leq 6$
##          $x_2 \geq 2$
##          $x_3 \geq 1$
##          $5x_1 + 2x_2 + 4x_3 - d_2^+ + d_2^- = 60$
##          $5x_1 + 5x_2 + 8x_3 - d_3^+ \leq 92$
##          $x_1, x_2, x_3 \geq 0$
##          $d_2^-, d_2^+, d_3^+ \geq 0$

In [55]:
import pandas as pd
import numpy as np
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, LpVariable
from pulp import GLPK

#
# Create the model
model_ppn = LpProblem(name="Projeto_individual_1_e_pn", sense= LpMinimize)

#
# Initialize the decision variables
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1,4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(2,3)}
dM = {i: LpVariable(name=f"dM{i}", lowBound=0) for i in range(2,4)}

vector = np.array([60,55])
peso = np.array([5,3,2])

#
# Add the constraints to the model
model_ppn += (x[1] <= 6, "produção D1")
model_ppn += (x[2] >= 2, "produção D2")
model_ppn += (x[3] >= 1, "produção D3")
model_ppn += (5 * x[1] + 2 * x[2] + 4 * x[3] - dM[2] + dm[2] == 60, "Mão de obra")
model_ppn += (5 * x[1] + 5 * x[2] + 8 * x[3] - dM[3] <= 92, "Investimento")
#
# Add the objective function to the model
obj_func = (peso[0]/vector[0]) * dm[2] + (peso[1]/vector[0]) * dM[2] + (peso[2]/vector[1]) * dM[3]
model_ppn += obj_func

# 
# Formulação (visualização do modelo matemático de forma a conferir os dados)
model_ppn

Projeto_individual_1_e_pn:
MINIMIZE
0.05*dM2 + 0.03636363636363636*dM3 + 0.08333333333333333*dm2 + 0.0
SUBJECT TO
produção_D1: x1 <= 6

produção_D2: x2 >= 2

produção_D3: x3 >= 1

Mão_de_obra: - dM2 + dm2 + 5 x1 + 2 x2 + 4 x3 = 60

Investimento: - dM3 + 5 x1 + 5 x2 + 8 x3 <= 92

VARIABLES
dM2 Continuous
dM3 Continuous
dm2 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [56]:
# Solve the problem
#
# Consideram-se três formas de resolver um problema de Programação Linear: 
# Forma 1: status = model.solve() 
# Forma 2: status = model.solve(solver=GLPK(msg=True))
# Forma 3: status = model.solve(GLPK(msg=True, options=['--ranges', 'sensitivity.txt']))
# A Forma 3 gera o relatório de análise de sensibilidade.
#
status = model_ppn.solve()

#
# Outputs
#
# Status (LpStatusOptimal=1; LpStatusNotSolved=0; LpStatusInfeasible=-1; LpStatusUnbounded=-2; LpStatusUndefined=-3)
#
# Valor óptimo de problema
model_ppn.objective.value()
print(f"objective: {model_ppn.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
for var in dM.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das variáveis de desvio
#
for name, constraint in model_ppn.constraints.items():
    print (f"{name}: {constraint.value()}") 

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/marcoesperanca/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/afece5be4a024031ab03f47e738a1ebf-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/afece5be4a024031ab03f47e738a1ebf-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 10 COLUMNS
At line 26 RHS
At line 32 BOUNDS
At line 33 ENDATA
Problem MODEL has 5 rows, 6 columns and 12 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 2 (-3) rows, 6 (0) columns and 9 (-3) elements
0  Obj 0 Primal inf 10.399999 (1)
2  Obj 5.2e-18
Optimal - objective value 0
After Postsolve, objective 0, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 0 - 2 iterations time 0.002, Presolve 0.00
Option for printingOptions changed from nor

### Segundo nível (primeira tentativa)

In [57]:
import pandas as pd
import numpy as np
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, LpVariable
from pulp import GLPK

#
# Create the model
model_psn = LpProblem(name="Projeto_individual_1_e_sn", sense= LpMinimize)

#
# Initialize the decision variables
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1,4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1,2)}


vector = np.array([125])
peso = np.array([5])


# Add the constraints to the model
model_psn += (x[1] <= 6, "produção D1")
model_psn += (x[2] >= 2, "produção D2")
model_psn += (x[3] >= 1, "produção D3")
model_psn += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1] >= 125, "Lucro a longo prazo")
model_psn += (5 * x[1] + 2 * x[2] + 4 * x[3]  == 60, "Mão de obra")
model_psn += (5 * x[1] + 5 * x[2] + 8 * x[3] <= 92, "Investimento")
#
# Add the objective function to the model
obj_func = (peso[0]/vector[0]) * dm[1]
model_psn += obj_func

# 
# Formulação (visualização do modelo matemático de forma a conferir os dados)
model_psn

Projeto_individual_1_e_sn:
MINIMIZE
0.04*dm1 + 0.0
SUBJECT TO
produção_D1: x1 <= 6

produção_D2: x2 >= 2

produção_D3: x3 >= 1

Lucro_a_longo_prazo: dm1 + 12 x1 + 9 x2 + 5 x3 >= 125

Mão_de_obra: 5 x1 + 2 x2 + 4 x3 = 60

Investimento: 5 x1 + 5 x2 + 8 x3 <= 92

VARIABLES
dm1 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [58]:
# Solve the problem
#
# Consideram-se três formas de resolver um problema de Programação Linear: 
# Forma 1: status = model.solve() 
# Forma 2: status = model.solve(solver=GLPK(msg=True))
# Forma 3: status = model.solve(GLPK(msg=True, options=['--ranges', 'sensitivity.txt']))
# A Forma 3 gera o relatório de análise de sensibilidade.
#
status = model_psn.solve()

#
# Outputs
#
# Status (LpStatusOptimal=1; LpStatusNotSolved=0; LpStatusInfeasible=-1; LpStatusUnbounded=-2; LpStatusUndefined=-3)
#
# Valor óptimo de problema
model_psn.objective.value()
print(f"objective: {model_psn.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das variáveis de desvio
#
for name, constraint in model_psn.constraints.items():
    print (f"{name}: {constraint.value()}") 

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/marcoesperanca/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/6f39ea03b86243dd9c3c4b3e61601ef0-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/6f39ea03b86243dd9c3c4b3e61601ef0-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 11 COLUMNS
At line 26 RHS
At line 33 BOUNDS
At line 34 ENDATA
Problem MODEL has 6 rows, 4 columns and 13 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 3 (-3) rows, 4 (0) columns and 10 (-3) elements
0  Obj 0 Primal inf 18.899998 (2)
4  Obj 0.1
Optimal - objective value 0.1
After Postsolve, objective 0.1, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 0.1 - 4 iterations time 0.002, Presolve 0.00
Option for printingOptions changed from 

### Segundo nível (segunda tentativa)

#### Reformulação do problema

## $\min P = P_1^-*\frac{d_1^{-}}{125}$


##   s.a. $x_1 \leq 6$
##          $x_2 \geq 2$
##          $x_3 \geq 1$
##          $12x_1 + 9x_2 + 5x_3 + d_1^-\geq 100$
##          $5x_1 + 2x_2 + 4x_3 = 60$
##          $5x_1 + 5x_2 + 8x_3  \leq 92$
##          $x_1, x_2, x_3 \geq 0$
##          $d_1^- \geq 0$

In [63]:
import pandas as pd
import numpy as np
from pulp import LpMaximize, LpMinimize, LpProblem, LpStatus, LpVariable
from pulp import GLPK

#
# Create the model
model_psn = LpProblem(name="Projeto_individual_1_e_sn", sense= LpMinimize)

#
# Initialize the decision variables
x = {i: LpVariable(name=f"x{i}", lowBound=0) for i in range(1,4)}
dm = {i: LpVariable(name=f"dm{i}", lowBound=0) for i in range(1,2)}


vector = np.array([125])
peso = np.array([5])


# Add the constraints to the model
model_psn += (x[1] <= 6, "produção D1")
model_psn += (x[2] >= 2, "produção D2")
model_psn += (x[3] >= 1, "produção D3")
model_psn += (12 * x[1] + 9 * x[2] + 5 * x[3] + dm[1] >= 100, "Lucro a longo prazo")
model_psn += (5 * x[1] + 2 * x[2] + 4 * x[3]  == 60, "Mão de obra")
model_psn += (5 * x[1] + 5 * x[2] + 8 * x[3] <= 92, "Investimento")
#
# Add the objective function to the model
obj_func = (peso[0]/vector[0]) * dm[1]
model_psn += obj_func

# 
# Formulação (visualização do modelo matemático de forma a conferir os dados)
model_psn

Projeto_individual_1_e_sn:
MINIMIZE
0.04*dm1 + 0.0
SUBJECT TO
produção_D1: x1 <= 6

produção_D2: x2 >= 2

produção_D3: x3 >= 1

Lucro_a_longo_prazo: dm1 + 12 x1 + 9 x2 + 5 x3 >= 100

Mão_de_obra: 5 x1 + 2 x2 + 4 x3 = 60

Investimento: 5 x1 + 5 x2 + 8 x3 <= 92

VARIABLES
dm1 Continuous
x1 Continuous
x2 Continuous
x3 Continuous

In [64]:
# Solve the problem
#
# Consideram-se três formas de resolver um problema de Programação Linear: 
# Forma 1: status = model.solve() 
# Forma 2: status = model.solve(solver=GLPK(msg=True))
# Forma 3: status = model.solve(GLPK(msg=True, options=['--ranges', 'sensitivity.txt']))
# A Forma 3 gera o relatório de análise de sensibilidade.
#
status = model_psn.solve()

#
# Outputs
#
# Status (LpStatusOptimal=1; LpStatusNotSolved=0; LpStatusInfeasible=-1; LpStatusUnbounded=-2; LpStatusUndefined=-3)
#
# Valor óptimo de problema
model_psn.objective.value()
print(f"objective: {model_psn.objective.value()}")
#
# Solução óptima
# Valores óptimos das variáveis de decisão
#
for var in x.values():
    print(f"{var.name}: {var.value()}")
for var in dm.values():
    print(f"{var.name}: {var.value()}")
#
# Valores das variáveis de desvio
#
for name, constraint in model_psn.constraints.items():
    print (f"{name}: {constraint.value()}") 

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/marcoesperanca/opt/anaconda3/lib/python3.9/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/16bfa1b6b7bf4ec48bf440555784dd02-pulp.mps timeMode elapsed branch printingOptions all solution /var/folders/6f/xxdlxkl953v7h6b6nsn7dhp40000gn/T/16bfa1b6b7bf4ec48bf440555784dd02-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 11 COLUMNS
At line 26 RHS
At line 33 BOUNDS
At line 34 ENDATA
Problem MODEL has 6 rows, 4 columns and 13 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 3 (-3) rows, 4 (0) columns and 10 (-3) elements
0  Obj 0 Primal inf 16.816665 (2)
2  Obj 5.2e-18
Optimal - objective value 0
After Postsolve, objective 0, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 0 - 2 iterations time 0.002, Presolve 0.00
Option for printingOptions changed from no